hil/ext/switches/ovs.py

```diff
@@ -0,0 +1,248 @@
+"""A switch driver for OpenVswitch."""
+import re
+import logging
+import schema
+import ast
+import subprocess
+import shlex
+
+from hil.model import db, Switch, BigIntegerType, SwitchSession
+from hil.errors import SwitchError
+
+logger = logging.getLogger(__name__)
+
+# Class layout
+# 1. Public methods:
+# 2. Private methods
+# 3. Unimplemented superclass methods.
+
+
+class Ovs(Switch, SwitchSession):
+    """ Driver for openvswitch. """
+    api_name = 'http://schema.massopencloud.org/haas/v0/switches/ovs'
+    __mapper_args__ = {
+        'polymorphic_identity': api_name,
+    }
+
+    id = db.Column(
+            BigIntegerType, db.ForeignKey('switch.id'), primary_key=True
+            )
+    hostname = db.Column(db.String, nullable=False)
+
+    @staticmethod
+    def validate(kwargs):
+        """Checks input to match switch parameters."""
+        schema.Schema({
+            'hostname': basestring,
+        }).validate(kwargs)
+
+# 1. Public Methods:
+
+    def get_capabilities(self):
+        return ['nativeless-trunk-mode']
+
+    def ovs_connect(self, *command_strings):
+        """Interacts with the Openvswitch.
+
+        Args:
+            *command_strings (tuple) : tuple of shell commands required
+                        to make changes to openvswitch
+        Raises: SwitchError
+        Returns: If successful returns None else logs error message
+        """
+        try:
+            for command in command_strings:
+                arg_list = shlex.split(command)
+                subprocess.check_call(arg_list)
+        except subprocess.CalledProcessError as e:
+            logger.error('%s', e)
+            raise SwitchError('ovs command failed: ')
+
+    def get_port_networks(self, ports):
+
+        response = {}
+        for port in ports:
+            trunk_id_list = self._interface_info(port.label)['trunks']
+            if trunk_id_list == []:
+                response[port] = []
+            else:
+                result = []
+                for trunk in trunk_id_list:
+                    name = "vlan/"+trunk
+                    result.append((name, trunk))
+                response[port] = result
+            native = self._interface_info(port.label)['tag']
+            if native != []:
+                response[port].append(("vlan/native", native))
+
+        return response
+
+    def revert_port(self, port):
+
+        shell_cmd_1 = 'sudo ovs-vsctl del-port {port}'.format(port=port)
+        shell_cmd_2 = 'sudo ovs-vsctl add-port {switch} {port}'.format(
+                    switch=self.hostname, port=port
+                    ) + ' vlan_mode=native-untagged'
+        self.ovs_connect(shell_cmd_1, shell_cmd_2)
+
+    def modify_port(self, port, channel, new_network):
+
+        (port,) = filter(lambda p: p.label == port, self.ports)
+        interface = port.label
+
+        if channel == 'vlan/native':
```

Comment:
I do not understand. Please elaborate.
This is the modify_port function. The logical branching is same as any other driver only the switch specific actions differ.

Reply:
I think it would be better to do the refactoring in a different PR if you guys don't feel too strongly about it.

Reply:
Ok, we do actually already have this copied in a bunch of other places.

I'm ok with deferring the refactoring until after merging this **if** someone is willing to take responsibility for doing the refactoring; I want to make sure we actually triage this rather than just letting it get worse and worse.